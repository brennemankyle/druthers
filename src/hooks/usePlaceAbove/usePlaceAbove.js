// eslint-disable-next-line
import React, { useState, useEffect } from 'react'

let usePlaceAbove = (rect, optionsWrapperRect, skip = []) => {
  const [placeOptionsAbove, setPlaceOptionsAbove] = useState(false)

  // if no space below, place above
  useEffect(() => {
    let windowHeight = window.innerHeight + window.scrollY
    let belowDiff = windowHeight - (rect.y + rect.height + optionsWrapperRect.height)
    let aboveDiff = rect.y - optionsWrapperRect.height
    setPlaceOptionsAbove(belowDiff < 0 && aboveDiff > belowDiff)

    // eslint-disable-next-line
  }, [rect, optionsWrapperRect].concat(skip))

  return placeOptionsAbove
}

export default usePlaceAbove
