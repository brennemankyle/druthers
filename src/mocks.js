let mockStyles = {
  colors: {}
}

let mockItem = {
  value: '42',
  label: 'Meaning of life',
}

let mockItemList = [
  {value: '1', label: 'Item 1'},
  {value: '2', label: 'Item 2'},
  {value: '3', label: 'Item 3'},
  {value: '4', label: 'Item 4'},
  {value: '5', label: 'Item 5'},
]

export { mockStyles, mockItem, mockItemList }
