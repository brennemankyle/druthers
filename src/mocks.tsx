import React from "react";
import defaultProps from "./utils/defaultProps";
import { withKeys, noop } from "./utils/utils";

let mockStyles = {
  styles_width: 600,
  styles_multiple: false,
  styles_disabled: false,
  styles_hasSelection: false,
  styles_hasOptions: true,
  styles_optionHighlighted: [],
  styles_rightToLeft: false,
  ...withKeys(defaultProps, "styles_"),
};

let mockItem = {
  value: "42",
  label: "Meaning of life",
};

let mockItemList = [
  { value: "1", label: "Item 1" },
  { value: "2", label: "Item 2" },
  { value: "3", label: "Item 3" },
  { value: "4", label: "Item 4" },
  { value: "5", label: "Item 5" },
];

let mockEvent = {
  preventDefault: noop,
};

let MockElement = (props: any) => <div data-props={props} />;
let MockInput = (props: any) => <input type="text" data-props={props} />;
let MockSvg = () => (
  <svg>
    <path d="M0 0L5 5Z" />
  </svg>
);

let mockRect = {
  top: 10,
  bottom: 110,
  left: 20,
  right: 220,
  width: 200,
  height: 100,
  x: 15,
  y: 25,
};

export {
  mockStyles,
  mockItem,
  mockItemList,
  mockEvent,
  MockElement,
  MockInput,
  MockSvg,
  mockRect,
};
