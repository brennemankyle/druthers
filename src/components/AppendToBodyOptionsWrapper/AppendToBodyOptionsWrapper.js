import React, { useRef, forwardRef } from 'react'
import PropTypes from 'prop-types'
import AppPropTypes from '../../utils/AppPropTypes'
import useRefRect from '../../hooks/useRefRect/useRefRect'
import usePlaceAbove from '../../hooks/usePlaceAbove/usePlaceAbove'
import Wrapper from '../Wrapper/Wrapper'

let StyledAppendToBodyOptionsWrapper = (props, ref) => {
  let {className, children, placeOptionsAbove, parentRect, ...otherProps} = props

  return <Wrapper className={className} ref={ref} {...otherProps}>{children}</Wrapper>
}

StyledAppendToBodyOptionsWrapper = forwardRef(StyledAppendToBodyOptionsWrapper)

let AppendToBodyOptionsWrapper = (props) => {
  let {className, children, parentRef, filteredOptions, updateOn, StyledAppendToBodyOptionsWrapper, ...otherProps} = props
  const ref = useRef(null)
  const parentRect = useRefRect(parentRef, updateOn)
  const optionsWrapperRect = useRefRect(ref, updateOn)
  const placeOptionsAbove = usePlaceAbove(parentRect, optionsWrapperRect, updateOn)

  if (placeOptionsAbove && children) {
    filteredOptions = filteredOptions.reverse()
    children = React.cloneElement(children, {itemList: filteredOptions})
  }

  return <StyledAppendToBodyOptionsWrapper
    className={className}
    ref={ref}
    placeOptionsAbove={placeOptionsAbove}
    parentRect={parentRect}
    {...otherProps}>{children}</StyledAppendToBodyOptionsWrapper>
}

AppendToBodyOptionsWrapper.defaultProps = {
  updateOn: [],
}

AppendToBodyOptionsWrapper.propTypes = {
  parentRef: PropTypes.object.isRequired,
  filteredOptions: AppPropTypes.itemList.isRequired,
  updateOn: PropTypes.array.isRequired,
  StyledAppendToBodyOptionsWrapper: AppPropTypes.element.isRequired,
  ...AppPropTypes.styles,
}

export { StyledAppendToBodyOptionsWrapper }
export default AppendToBodyOptionsWrapper
