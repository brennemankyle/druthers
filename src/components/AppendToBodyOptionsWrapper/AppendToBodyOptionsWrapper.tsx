import React, {
  useRef,
  forwardRef,
  ReactElement,
  ForwardedRef,
  ForwardRefExoticComponent,
  RefAttributes,
} from "react";
import useRefRect from "../../hooks/useRefRect/useRefRect";
import usePlaceAbove from "../../hooks/usePlaceAbove/usePlaceAbove";
import Wrapper from "../Wrapper/Wrapper";
import { Item, Styles } from "../../utils/SelectTypes";
import { DomRect } from "../../hooks/useRefRect/useRefRect";

export interface StyledProps {
  className: string;
  children: ReactElement;
  placeOptionsAbove: boolean;
  parentRect: DomRect;
}

export type StyledAppendToBodyOptionsWrapperProps = StyledProps & Styles;

// Only props can by used for styled components, therefore AppendToBodyOptionsWrapper can't use it's state to be styled
const StyledAppendToBodyOptionsWrapper = forwardRef(
  function StyledAppendToBodyOptionsWrapper(
    props: StyledAppendToBodyOptionsWrapperProps,
    ref: ForwardedRef<HTMLDivElement> | undefined
  ) {
    let { className, children, placeOptionsAbove, parentRect, ...otherProps } =
      props;

    return (
      <Wrapper className={className} ref={ref} {...otherProps}>
        {children}
      </Wrapper>
    );
  }
);

export interface Props {
  className: string;
  children: ReactElement;
  parentRef: null;
  filteredOptions: Item[];
  updateOn: string[];
  StyledWrapper: ForwardRefExoticComponent<
    StyledAppendToBodyOptionsWrapperProps & RefAttributes<HTMLDivElement>
  >;
}

export type AppendToBodyOptionsWrapperProps = Styles & Props;

function AppendToBodyOptionsWrapper({
  className,
  children,
  parentRef,
  filteredOptions,
  updateOn = [],
  StyledWrapper = StyledAppendToBodyOptionsWrapper,
  ...otherProps
}: AppendToBodyOptionsWrapperProps) {
  const ref = useRef(null);
  const parentRect = useRefRect(parentRef, updateOn);
  const optionsWrapperRect = useRefRect(ref, updateOn);
  const placeOptionsAbove = usePlaceAbove(
    parentRect,
    optionsWrapperRect,
    updateOn
  );

  if (placeOptionsAbove && children) {
    filteredOptions = filteredOptions.reverse();
    children = React.cloneElement(children, { itemList: filteredOptions });
  }

  return (
    <StyledWrapper
      className={className}
      ref={ref}
      placeOptionsAbove={placeOptionsAbove}
      parentRect={parentRect}
      {...otherProps}
    >
      {children}
    </StyledWrapper>
  );
}

export { StyledAppendToBodyOptionsWrapper };
export default AppendToBodyOptionsWrapper;
