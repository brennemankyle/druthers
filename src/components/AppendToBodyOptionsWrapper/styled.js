import styled from "@emotion/styled";
import { StyledAppendToBodyOptionsWrapper as RawStyledAppendToBodyOptionsWrapper } from "./AppendToBodyOptionsWrapper";

const StyledAppendToBodyOptionsWrapper = styled(
  RawStyledAppendToBodyOptionsWrapper
)`
  position: absolute;
  background-color: ${props => props.styles_colors_background};
  width: ${props => props.styles_width}px;

  ${props =>
    props.placeOptionsAbove
      ? `bottom: ${window.innerHeight - props.parentRect.y}px;
      left: ${props.parentRect.x}px;`
      : `top: ${props.parentRect.y + props.parentRect.height}px;
      left: ${props.parentRect.x}px;`}
  ${props => props.styles_rightToLeft && `transform: scaleX(-1);`}
  ${props =>
    !props.styles_hasOptions && !props.styles_multiple && `display: none;`}
`;

export { StyledAppendToBodyOptionsWrapper };
