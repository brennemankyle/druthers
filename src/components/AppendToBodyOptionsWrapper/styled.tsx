import React from "react";
import styled from "@emotion/styled";
import {
  StyledAppendToBodyOptionsWrapper as RawStyledAppendToBodyOptionsWrapper,
  StyledAppendToBodyOptionsWrapperProps,
} from "./AppendToBodyOptionsWrapper";
import { Styles } from "../../utils/SelectTypes";
import { createBaseOptionsWrapper } from "../Wrapper/styled";

const BaseOptionsWrapper = createBaseOptionsWrapper(
  RawStyledAppendToBodyOptionsWrapper
);

const StyledAppendToBodyOptionsWrapper = styled(BaseOptionsWrapper)(
  (props: Styles & StyledAppendToBodyOptionsWrapperProps) => `
  position: absolute;
  background-color: ${props.styles_colors_background};

  ${
    props.placeOptionsAbove
      ? `bottom: ${window.innerHeight - props.parentRect.y}px;
        left: ${props.parentRect.x}px;`
      : `top: ${props.parentRect.y + props.parentRect.height}px;
        left: ${props.parentRect.x}px;`
  }
  ${props.styles_rightToLeft ? `transform: scaleX(-1);` : ``}
  ${!props.styles_hasOptions && !props.styles_multiple ? `display: none;` : ``}
`
);

export { StyledAppendToBodyOptionsWrapper };
