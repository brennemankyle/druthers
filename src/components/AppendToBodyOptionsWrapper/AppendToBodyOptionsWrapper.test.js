import React from "react";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import usePlaceAbove from "../../hooks/usePlaceAbove/usePlaceAbove";
import { mockStyles, MockElement, mockItemList } from "../../mocks";
import AppendToBodyOptionsWrapper from "./AppendToBodyOptionsWrapper";

it("renders", () => {
  const parentRef = React.createRef();
  const wrapper = mount(
    <AppendToBodyOptionsWrapper
      {...mockStyles}
      className="test"
      parentRef={parentRef}
      filteredOptions={mockItemList}
      StyledAppendToBodyOptionsWrapper={MockElement}
    >
      <MockElement />
    </AppendToBodyOptionsWrapper>
  );

  expect(toJson(wrapper)).toMatchSnapshot();
});

it("forwards ref", () => {
  const parentRef = React.createRef();
  shallow(
    <AppendToBodyOptionsWrapper
      {...mockStyles}
      className="test"
      parentRef={parentRef}
      filteredOptions={[]}
      StyledAppendToBodyOptionsWrapper={MockElement}
    />
  );

  expect(parentRef.current).toBeDefined();
});

it("place above", () => {
  usePlaceAbove.mockImplementation(() => true);

  const parentRef = React.createRef();
  const wrapper = mount(
    <AppendToBodyOptionsWrapper
      {...mockStyles}
      className="test"
      parentRef={parentRef}
      filteredOptions={mockItemList}
      StyledAppendToBodyOptionsWrapper={MockElement}
    >
      <MockElement />
    </AppendToBodyOptionsWrapper>
  );

  expect(toJson(wrapper)).toMatchSnapshot();
});
