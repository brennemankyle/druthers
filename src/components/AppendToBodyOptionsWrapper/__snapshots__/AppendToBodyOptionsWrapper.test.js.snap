// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<AppendToBodyOptionsWrapper
  StyledAppendToBodyOptionsWrapper={[Function]}
  className="test"
  filteredOptions={
    Array [
      Object {
        "label": "Item 5",
        "value": "5",
      },
      Object {
        "label": "Item 4",
        "value": "4",
      },
      Object {
        "label": "Item 3",
        "value": "3",
      },
      Object {
        "label": "Item 2",
        "value": "2",
      },
      Object {
        "label": "Item 1",
        "value": "1",
      },
    ]
  }
  parentRef={
    Object {
      "current": null,
    }
  }
  styles_borderRadius=".2em"
  styles_checkRadio_borderWidth="2px"
  styles_checkRadio_labelMargin=".2em"
  styles_checkRadio_marginBetween="1em"
  styles_checkRadio_paddingBottom=".2em"
  styles_checkRadio_paddingLeft=".2em"
  styles_checkRadio_paddingRight=".2em"
  styles_checkRadio_paddingTop=".2em"
  styles_colors_background="white"
  styles_colors_disabled="#ECECEC"
  styles_colors_highlight="lightblue"
  styles_colors_primary="black"
  styles_colors_secondary="#C3C3C3"
  styles_colors_warning="#FABAAC"
  styles_colors_warningBold="#FA2222"
  styles_disabled={false}
  styles_fontSize="1em"
  styles_hasOptions={true}
  styles_hasSelection={false}
  styles_icon_width="1em"
  styles_multiple={false}
  styles_optionHighlighted={Array []}
  styles_option_paddingBottom=".25em"
  styles_option_paddingLeft=".25em"
  styles_option_paddingRight=".25em"
  styles_option_paddingTop=".25em"
  styles_paddingBottom=".25em"
  styles_paddingLeft=".25em"
  styles_paddingRight=".25em"
  styles_paddingTop=".25em"
  styles_rightToLeft={false}
  styles_search_size="8"
  styles_selection_margin=".2em"
  styles_selection_paddingBottom=".2em"
  styles_selection_paddingLeft=".4em"
  styles_selection_paddingRight=".4em"
  styles_selection_paddingTop=".2em"
  styles_width="600px"
  updateOn={Array []}
>
  <MockElement
    className="test"
    parentRect={
      Object {
        "bottom": 80,
        "height": 80,
        "left": 0,
        "right": 200,
        "top": 0,
        "width": 200,
        "x": 0,
        "y": 600,
      }
    }
    placeOptionsAbove={true}
    styles_borderRadius=".2em"
    styles_checkRadio_borderWidth="2px"
    styles_checkRadio_labelMargin=".2em"
    styles_checkRadio_marginBetween="1em"
    styles_checkRadio_paddingBottom=".2em"
    styles_checkRadio_paddingLeft=".2em"
    styles_checkRadio_paddingRight=".2em"
    styles_checkRadio_paddingTop=".2em"
    styles_colors_background="white"
    styles_colors_disabled="#ECECEC"
    styles_colors_highlight="lightblue"
    styles_colors_primary="black"
    styles_colors_secondary="#C3C3C3"
    styles_colors_warning="#FABAAC"
    styles_colors_warningBold="#FA2222"
    styles_disabled={false}
    styles_fontSize="1em"
    styles_hasOptions={true}
    styles_hasSelection={false}
    styles_icon_width="1em"
    styles_multiple={false}
    styles_optionHighlighted={Array []}
    styles_option_paddingBottom=".25em"
    styles_option_paddingLeft=".25em"
    styles_option_paddingRight=".25em"
    styles_option_paddingTop=".25em"
    styles_paddingBottom=".25em"
    styles_paddingLeft=".25em"
    styles_paddingRight=".25em"
    styles_paddingTop=".25em"
    styles_rightToLeft={false}
    styles_search_size="8"
    styles_selection_margin=".2em"
    styles_selection_paddingBottom=".2em"
    styles_selection_paddingLeft=".4em"
    styles_selection_paddingRight=".4em"
    styles_selection_paddingTop=".2em"
    styles_width="600px"
  >
    <div
      data-props={
        Object {
          "children": <MockElement
            itemList={
              Array [
                Object {
                  "label": "Item 5",
                  "value": "5",
                },
                Object {
                  "label": "Item 4",
                  "value": "4",
                },
                Object {
                  "label": "Item 3",
                  "value": "3",
                },
                Object {
                  "label": "Item 2",
                  "value": "2",
                },
                Object {
                  "label": "Item 1",
                  "value": "1",
                },
              ]
            }
          />,
          "className": "test",
          "parentRect": Object {
            "bottom": 80,
            "height": 80,
            "left": 0,
            "right": 200,
            "top": 0,
            "width": 200,
            "x": 0,
            "y": 600,
          },
          "placeOptionsAbove": true,
          "styles_borderRadius": ".2em",
          "styles_checkRadio_borderWidth": "2px",
          "styles_checkRadio_labelMargin": ".2em",
          "styles_checkRadio_marginBetween": "1em",
          "styles_checkRadio_paddingBottom": ".2em",
          "styles_checkRadio_paddingLeft": ".2em",
          "styles_checkRadio_paddingRight": ".2em",
          "styles_checkRadio_paddingTop": ".2em",
          "styles_colors_background": "white",
          "styles_colors_disabled": "#ECECEC",
          "styles_colors_highlight": "lightblue",
          "styles_colors_primary": "black",
          "styles_colors_secondary": "#C3C3C3",
          "styles_colors_warning": "#FABAAC",
          "styles_colors_warningBold": "#FA2222",
          "styles_disabled": false,
          "styles_fontSize": "1em",
          "styles_hasOptions": true,
          "styles_hasSelection": false,
          "styles_icon_width": "1em",
          "styles_multiple": false,
          "styles_optionHighlighted": Array [],
          "styles_option_paddingBottom": ".25em",
          "styles_option_paddingLeft": ".25em",
          "styles_option_paddingRight": ".25em",
          "styles_option_paddingTop": ".25em",
          "styles_paddingBottom": ".25em",
          "styles_paddingLeft": ".25em",
          "styles_paddingRight": ".25em",
          "styles_paddingTop": ".25em",
          "styles_rightToLeft": false,
          "styles_search_size": "8",
          "styles_selection_margin": ".2em",
          "styles_selection_paddingBottom": ".2em",
          "styles_selection_paddingLeft": ".4em",
          "styles_selection_paddingRight": ".4em",
          "styles_selection_paddingTop": ".2em",
          "styles_width": "600px",
        }
      }
    />
  </MockElement>
</AppendToBodyOptionsWrapper>
`;
